Case Study on Regression
The given dataset contains price of second-hand Hyundai grand i10 car with 
respect to year of making. Find the best linear relationship between year and 
price. Can you predict the possible price of a 2022 model second-hand grand i10?
Please learn about lasso regression and create a model along with linear
regression. Find out which one is performing better.

# Import Necessary Libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Load the Dataset


In [2]:
data = pd.read_csv(r'C:\Users\LINI\Downloads\car_age_price.csv')

In [3]:
data.head()

Year   Price
0  2018  465000
1  2019  755000
2  2019  700000
3  2018  465000
4  2018  465000

In [4]:
data.shape

(112, 2)

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112 entries, 0 to 111
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   Year    112 non-null    int64
 1   Price   112 non-null    int64
dtypes: int64(2)
memory usage: 1.9 KB


In [6]:
data.describe()

Year          Price
count   112.000000     112.000000
mean   2016.669643  483866.044643
std       1.629616   91217.450533
min    2013.000000  300000.000000
25%    2015.000000  423750.000000
50%    2017.000000  500000.000000
75%    2017.000000  550000.000000
max    2020.000000  755000.000000

In [7]:
data.isna().sum()

Year     0
Price    0
dtype: int64

# Linear Regression

In [8]:
X=data.drop(['Price'],axis=1)
y=data['Price']

In [9]:
from sklearn.model_selection import train_test_split


Then split the data into training and testing set

In [10]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size = 0.25,random_state = 30)

In [11]:
from sklearn.linear_model import LinearRegression

In [12]:
lr = LinearRegression()

In [13]:
Linear_model = lr.fit(X_train,y_train)

In [14]:
y_lr_pred = Linear_model.predict(X_test)

In [15]:
from sklearn.metrics import mean_squared_error,r2_score


In [16]:
linear_mse = mean_squared_error(y_test,y_lr_pred)
linear_r2  = r2_score(y_test,y_lr_pred)

# Lasso Regression

In [17]:
from sklearn.linear_model import Lasso

In [18]:
lasso = Lasso(alpha = 0.1) #alpha is the regularization Parametre

In [19]:
Lasso_model = lasso.fit(X_train,y_train)

In [20]:
y_lasso_pred = Lasso_model.predict(X_test)

In [21]:
lasso_mse = mean_squared_error(y_test,y_lasso_pred)
lasso_r2  = r2_score(y_test,y_lasso_pred)

In [22]:
print("Linear MSE :",linear_mse)
print("Lasso MSE :",lasso_mse)
print("Linear R2 :",linear_r2)
print("Lasso R2:",lasso_r2)

Linear MSE : 3323580279.3057265
Lasso MSE : 3323578497.8480916
Linear R2 : 0.5492511227236985
Lasso R2: 0.54925136432762


In [23]:
if linear_mse < lasso_mse and linear_r2 > lasso_r2:
    selected_model = Linear_model
    print("Selected model : Linear regression")
elif lasso_mse < linear_mse and lasso_r2 > linear_r2:
    selected_model = Lasso_model
    print("Selected model : Lasso regression ")
else:
    print("Both models have similar performance.") 


Selected model : Lasso regression 


In [24]:
predicted_price_2022 = selected_model.predict([[2022]])

C:\Users\LINI\anaconda3\lib\site-packages\sklearn\base.py:420: UserWarning: X does not have valid feature names, but Lasso was fitted with feature names
  warnings.warn(


In [25]:
predicted_price_2022

array([721607.06995736])

# INSIGHTS : We can find that Lasso Model has better efficiency than Linear model and the predicted  price of 2022 model second-hand grand i10 using Lasso Model is 721607.06995736